### PR TITLE
Add tests for HTML encoding of report names

### PR DESCRIPTION
**Pull request by labkey-sweta**
#### Rationale
Automated coverage for report name HTML encoding

Issue : https://www.labkey.org/home/Developer/issues/Secure/issues-details.view?issueId=40558

#### Related Pull Requests
Only this one.

#### Changes
1. Added new test class to DailyB
2. New method in ManageViewPage.

### CODE REVIEW

File: src/org/labkey/test/pages/reports/ManageViewsPage.java [hunk 1]
```diff
@@ -63,4 +63,12 @@ public void deleteReport(String reportName)
                 "Failed to delete report: " + reportName, WAIT_FOR_JAVASCRIPT);
     }
 
+    @LogMethod
+    public void editReport(String reportName)
+    {
+        //td[1] is the pencil
+        final Locator report = Locator.tag("tr").withClass("x4-grid-row").containing(reportName).childTag("td").position(1);
+        click(report);
```
Comment (labkey-tchad) [inline src/org/labkey/test/pages/reports/ManageViewsPage.java:71]: If this navigates, we should wait for the page load.
```suggestion
        clickAndWait(report);
```

File: src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Test
+    public void verifyReportNameIsEncoded()
+    {
+        goToProjectHome();
+        goToManageViews();
+
+        log("Adding the link report");
+        new BootstrapMenu.BootstrapMenuFinder(getDriver()).withButtonTextContaining("Add Report").find()
+                .clickSubMenu(true, "Link Report");
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java]: Use the page returned by `goToManageViews`.
```suggestion
        log("Adding the link report");
        goToManageViews().clickAddReport("Link Report");
```

File: src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.util.Collections;
+import java.util.List;
+
+@Category({DailyB.class})
+@BaseWebDriverTest.ClassTimeout(minutes = 3)
+public class CrossSiteScriptingForDeleteTest extends BaseWebDriverTest
+{
+    protected static final String PROJECT_NAME = "CrossSiteScriptingForDeleteTest";
+    protected static final String REPORT_NAME = "<span onclick='alert(\"fooled you!\")'>mouseover me, it's totally safe</span>";
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java]: Use an injection string that doesn't require direct interaction. It gives us a chance to catch something we aren't specifically checking for.
```suggestion
    protected static final String REPORT_NAME = BaseWebDriverTest.INJECT_CHARS_1;
```

File: src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                .clickSubMenu(true, "Link Report");
+
+        setFormElement(Locator.name("viewName"), REPORT_NAME);
+        setFormElement(Locator.name("linkUrl"), WebTestHelper.getContextPath() + LINK_REPORT_URL);
+        clickButton("Save");
+        waitForText("Manage Views");
+
+        log("Clicking on the report - No XSS");
+        click(Locator.linkWithText(REPORT_NAME));
+        switchToMainWindow();
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java:68]: Add a `selectReport` method to `ManageViewsPage`. Check `ManageViewsPage.deleteReport` for a Locator that won't open the report.

**Review by labkey-tchad — CHANGES_REQUESTED** [review]

File: src/org/labkey/test/pages/reports/ManageViewsPage.java (hunk 1 above)
Comment (labkey-sweta) [reply src/org/labkey/test/pages/reports/ManageViewsPage.java:71]: It does not navigate, that's why only click.

File: src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java (hunk 2 above)
Comment (labkey-sweta) [reply src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java]: Done

File: src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java (hunk 4 above)
Comment (labkey-sweta) [reply src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java:68]: Done

File: src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java (hunk 3 above)
Comment (labkey-sweta) [reply src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java]: Done.

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]
This looks like it should prevent regressions on the two delete dialogs that had problems, but do nothing to help us detect any other XSS with report names. There are plenty of other places in the code that could have problems, like we saw with the user DisplayName property. It was only when we started setting the display name for the user our tests were running as that we discovered those other places.

I don't expect the automation changes here to protect against any and all possible report-related XSS, but can we please go more general than this? Lots of other tests create reports. We could update some of them to set report names, descriptions, and other values that verify that we don't have XSS issues.

I think that @labkey-tchad  and @cnathe  both worked on the user display name testing and fixes. Hopefully they will have some suggestions here too.

**Review by labkey-tchad — APPROVED** [review]

**Comment by labkey-tchad** [comment]
I think a good target would be to add injection characters to all of the report names, titles, and descriptions in `DataRerportsTest`, `NonStudyReportsTest`, and `ScatterPlotTest` (or any other `GenericChartsTest` subclasses).

File: src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                .waitFor()
+                .clickButton("Delete", false);
+        deleteMsg = "Are you sure you want to delete \"" + REPORT_NAME + "\"";
+
+        log("Verifying the error message");
+        confirmReportNameInMessage("Delete?", deleteMsg, true);
+
+    }
+
+    public void confirmReportNameInMessage(String windowTitle, String msg, boolean action)
```
Comment (labkey-danield) [inline src/org/labkey/test/tests/CrossSiteScriptingForDeleteTest.java:88]: Personally I would change the name of this to something like confirmReportNameAndDoAction. The current name of the function implies it is just a check. Not a big deal.

You may also want to change the access to private, unless you expect it to be used outside of the test.

**Review by labkey-danield — APPROVED** [review]

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]
NonStudyReportsTest isn't passing in the feature branch. ExtHelper.Locators.extDialog() isn't properly escaping the report name when it builds the XPath, which causes the test to fail.

   org.openqa.selenium.InvalidSelectorException: invalid selector: Unable to locate an element with the xpath expression 

I can repro this locally.

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]
Instead of making the validation less rigorous by removing some characters, I'd really prefer that we properly encode the XPath. We must do this for at least some other Locators.

https://stackoverflow.com/questions/642125/encoding-xpath-expressions-with-both-single-and-double-quotes

**Comment by labkey-tchad** [comment]
> Instead of making the validation less rigorous by removing some characters, I'd really prefer that we properly encode the XPath. We must do this for at least some other Locators.
> 

@labkey-sweta `Locator.containing` and `Locator.withText` will escape text properly.

**Review by labkey-jeckels — APPROVED** [review]
DailyC is still running, but the test of interest, NonStudyReportsTest, has passed, so it should be safe to merge.

File: src/org/labkey/test/util/ExtHelper.java [hunk 6]
```diff
@@ -565,8 +565,9 @@ private static String jsString(String s)
          */
         public static Locator.XPathLocator extDialog(String title)
         {
-            return Locator.xpath("//span[" + Locator.NOT_HIDDEN + " and contains(@class, 'window-header-text') and contains(string(), '" + title + "')]");
-        }
+            return Locator.xpath("//div").withClass("x4-window").notHidden().
+                    withDescendant(Locator.xpath("//span").withClass("x4-window-header-text").withText(title));
+     }
```
Comment (labkey-tchad) [inline src/org/labkey/test/util/ExtHelper.java]: This Locator isn't intended to find the window, just the title bar. It also needs to work with ext3 and we shouldn't impose a strict text match where there wasn't one before.
```suggestion
            return Locator.tag("span").notHidden().withAttributeContaining("class", "window-header-text").containing(title);
         }
```

**Review by labkey-tchad — CHANGES_REQUESTED** [review]
`ExtHelper` refactor is a bit too aggressive.

**Review by labkey-tchad — APPROVED** [review]

**Pull request merged**